crates/sdk/src/masp.rs

```diff
@@ -586,142 +650,147 @@ impl<U: ShieldedUtils + MaybeSend + MaybeSync> ShieldedContext<U> {
-                        *epoch,
-                        changed_keys,
-                        stx,
-                        native_token.clone(),
-                    )?;
-                } else {
-                    break;
-                }
+            let mut vk_heights = BTreeMap::new();
+            std::mem::swap(&mut vk_heights, &mut self.vk_heights);
+            for (vk, h) in vk_heights
+                .iter_mut()
+                .filter(|(_vk, h)| **h < Some(indexed_tx))
+            {
+                self.scan_tx(
+                    indexed_tx,
+                    epoch,
+                    &tx,
+                    &stx,
+                    vk,
+                    native_token.clone(),
+                )?;
+                *h = Some(indexed_tx);
             }
-            // Merge the context data originating from the unknown keys into the
-            // current context
-            self.merge(tx_ctx);
-        } else {
-            // Load only transactions accepted from last_txid until this point
-            txs = Self::fetch_shielded_transfers(client, self.last_indexed)
-                .await?;
-            tx_iter = txs.iter();
-        }
-        // Now that we possess the unspent notes corresponding to both old and
-        // new keys up until tx_pos, proceed to scan the new transactions.
-        for (indexed_tx, (epoch, changed_keys, stx)) in &mut tx_iter {
-            self.scan_tx(
-                *indexed_tx,
-                *epoch,
-                changed_keys,
-                stx,
-                native_token.clone(),
-            )?;
+            // possibly remove unneeded elements from the cache.
+            self.unscanned.scanned(&indexed_tx);
+            std::mem::swap(&mut vk_heights, &mut self.vk_heights);
+            self.utils
+                .save(self)
+                .await
+                .map_err(|e| Error::Other(e.to_string()))?;
         }
+
         Ok(())
     }
 
     /// Obtain a chronologically-ordered list of all accepted shielded
     /// transactions from a node.
-    pub async fn fetch_shielded_transfers<C: Client + Sync>(
+    pub async fn fetch_shielded_transfers<C: Client + Sync, IO: Io>(
+        &self,
         client: &C,
-        last_indexed_tx: Option<IndexedTx>,
-    ) -> Result<
-        BTreeMap<
-            IndexedTx,
-            (
-                Epoch,
-                BTreeSet<namada_core::types::storage::Key>,
-                Transaction,
-            ),
-        >,
-        Error,
-    > {
+        logger: &impl ProgressLogger<IO>,
+        last_indexed_tx: Option<BlockHeight>,
+        last_query_height: Option<BlockHeight>,
+    ) -> Result<IndexedNoteData, Error> {
         // Query for the last produced block height
         let last_block_height = query_block(client)
             .await?
             .map_or_else(BlockHeight::first, |block| block.height);
+        let last_query_height = last_query_height.unwrap_or(last_block_height);
```

Comment:
Should we query for the block height only if the arg is `None`? Just to avoid an extra RPC call since we only use `last_block_height`. Anyway this is very minor, we could change it in another PR in case